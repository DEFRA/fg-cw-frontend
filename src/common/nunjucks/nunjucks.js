import Vision from "@hapi/vision";
import Nunjucks from "nunjucks";

import { glob } from "node:fs/promises";
import { config } from "../config.js";
import { filters } from "../format/format-filters.js";
import { context } from "./context/context.js";
import { formatDate } from "./filters/format-date.js";

const njkPaths = [
  "node_modules/govuk-frontend/dist/",
  ...(await Array.fromAsync(
    glob(["**/views", "**/*(layouts|components|pages|partials)/"], {
      exclude: ["node_modules", "*.*"],
    }),
  )),
];

const viewPaths = await Array.fromAsync(
  glob(["**/views"], {
    exclude: ["node_modules", "*.*"],
  }),
);

export const environment = Nunjucks.configure(njkPaths, {
  autoescape: true,
  throwOnUndefined: false,
  trimBlocks: true,
  lstripBlocks: true,
  watch: config.get("nunjucks.watch"),
  noCache: config.get("nunjucks.noCache"),
  filter: {
    formatDate,
  },
});

environment.addFilter("formatDate", formatDate);
environment.addFilter("fixed", filters.fixed);

export const nunjucks = {
  plugin: Vision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = Nunjucks.compile(src, options.environment);
          return (ctx) => template.render(ctx);
        },
      },
    },
    compileOptions: {
      environment,
    },
    relativeTo: config.get("root"),
    path: viewPaths,
    isCached: config.get("isProduction"),
    context,
  },
};
