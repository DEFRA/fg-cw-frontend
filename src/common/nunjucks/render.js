import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
