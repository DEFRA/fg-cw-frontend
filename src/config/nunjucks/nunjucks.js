import { fileURLToPath } from 'node:url'
import path from 'path'
import nunjucks from 'nunjucks'
import hapiVision from '@hapi/vision'

import { config } from '../config.js'
import { context } from './context/context.js'
import * as filters from './filters/filters.js'
import * as globals from './globals.js'

const dirname = path.dirname(fileURLToPath(import.meta.url))
const nunjucksEnvironment = nunjucks.configure(
  [
    'node_modules/govuk-frontend/dist/',
    path.resolve(dirname, '../../server/common/templates'),
    path.resolve(dirname, '../../server/common/components'),
    path.resolve(dirname, '../../case-management/views'),
    path.resolve(dirname, '../../case-management/views/layouts'),
    path.resolve(dirname, '../../case-management/views/components'),
    path.resolve(dirname, '../../case-management/views/pages'),
    path.resolve(dirname, '../../case-management/views/partials'),
    path.resolve(dirname, '../../common/presentation')
  ],
  {
    autoescape: true,
    throwOnUndefined: false,
    trimBlocks: true,
    lstripBlocks: true,
    watch: config.get('nunjucks.watch'),
    noCache: config.get('nunjucks.noCache')
  }
)

export const nunjucksConfig = {
  plugin: hapiVision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = nunjucks.compile(src, options.environment)
          return (ctx) => template.render(ctx)
        }
      }
    },
    compileOptions: {
      environment: nunjucksEnvironment
    },
    relativeTo: path.resolve(dirname, '../..'),
    path: ['server', 'case-management/views', 'common/presentation'],
    isCached: config.get('isProduction'),
    context
  }
}

Object.entries(globals).forEach(([name, global]) => {
  nunjucksEnvironment.addGlobal(name, global)
})

Object.entries(filters).forEach(([name, filter]) => {
  nunjucksEnvironment.addFilter(name, filter)
})
