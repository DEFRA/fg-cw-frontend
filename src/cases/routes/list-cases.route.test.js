import hapi from "@hapi/hapi";
import { load } from "cheerio";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { nunjucks } from "../../common/nunjucks/nunjucks.js";
import { findAllCasesUseCase } from "../use-cases/find-all-cases.use-case.js";
import { listCasesRoute } from "./list-cases.route.js";

vi.mock("../use-cases/find-all-cases.use-case.js");

describe("listCasesRoute", () => {
  let server;

  beforeAll(async () => {
    server = hapi.server();
    server.route(listCasesRoute);
    await server.register([nunjucks]);

    await server.initialize();
  });

  afterAll(async () => {
    await server.stop();
  });

  it("returns a list of cases", async () => {
    findAllCasesUseCase.mockResolvedValue(mockCases);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases",
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("extracts assignedCaseId from query parameters and renders notification banner", async () => {
    findAllCasesUseCase.mockResolvedValue(mockCases);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: `/cases?assignedCaseId=${mockCases[0]._id}`,
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $(".govuk-notification-banner--success").html();

    expect(view).toMatchSnapshot();
  });
});

const mockCases = [
  {
    _id: "68495db5afe2d27b09b2ee47",
    caseRef: "banana-123",
    workflowCode: "frps-private-beta",
    status: "NEW",
    dateReceived: "2025-06-11T10:43:01.603Z",
    currentStage: "contract",
    payload: {
      clientRef: "banana-123",
      code: "frps-private-beta",
      createdAt: "2025-06-11T10:43:01.417Z",
      submittedAt: "2023-10-01T12:00:00.000Z",
      identifiers: {
        sbi: "SBI001",
        frn: "FIRM0001",
        crn: "CUST0001",
        defraId: "DEFRA0001",
      },
      answers: {
        agreementName: "Test application name 1",
        scheme: "SFI",
        year: 2025,
        hasCheckedLandIsUpToDate: true,
        actionApplications: [
          {
            parcelId: "9238",
            sheetId: "SX0679",
            code: "CSAM1",
            appliedFor: {
              unit: "ha",
              quantity: 20.23,
            },
          },
        ],
      },
    },
    stages: [
      {
        id: "application-receipt",
        taskGroups: [
          {
            id: "application-receipt-tasks",
            tasks: [
              {
                id: "simple-review",
                isComplete: false,
              },
            ],
          },
        ],
      },
      {
        id: "contract",
        taskGroups: [],
      },
    ],
    assignedUser: {
      id: "user-id-1",
      name: "John Smith",
    },
  },
  {
    _id: "68495db5afe2d27b09b2ee99",
    caseRef: "case-ref-2",
    workflowCode: "frps-private-beta",
    status: "OLD",
    dateReceived: "2025-06-11T10:43:01.603Z",
    currentStage: "contract",
    payload: {
      clientRef: "banana-123",
      code: "frps-private-beta",
      createdAt: "2025-06-11T10:43:01.417Z",
      submittedAt: "2023-10-01T12:00:00.000Z",
      identifiers: {
        sbi: "SBI001",
        frn: "FIRM0001",
        crn: "CUST0001",
        defraId: "DEFRA0001",
      },
      answers: {
        agreementName: "Test application name 1",
        scheme: "SFI",
        year: 2025,
        hasCheckedLandIsUpToDate: true,
        actionApplications: [
          {
            parcelId: "9238",
            sheetId: "SX0679",
            code: "CSAM1",
            appliedFor: {
              unit: "ha",
              quantity: 20.23,
            },
          },
        ],
      },
    },
    stages: [
      {
        id: "application-receipt",
        taskGroups: [
          {
            id: "application-receipt-tasks",
            tasks: [
              {
                id: "simple-review",
                isComplete: false,
              },
            ],
          },
        ],
      },
      {
        id: "contract",
        taskGroups: [],
      },
    ],
  },
];
