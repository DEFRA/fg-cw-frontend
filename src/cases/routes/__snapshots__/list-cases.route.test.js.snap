// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`listCasesRoute > extracts assignedCaseId from query parameters and renders notification banner 1`] = `
"
  <div class="govuk-notification-banner__header">
    <h2 class="govuk-notification-banner__title" id="govuk-notification-banner-title">
      Success
    </h2>
  </div>
  <div class="govuk-notification-banner__content">
    <h3 class="govuk-notification-banner__heading">Case assigned successfully</h3>
          <p class="govuk-body">      <a class="govuk-notification-banner__link" href="/cases/68495db5afe2d27b09b2ee47">Case banana-123</a>
          has been assigned to
          John Smith
    </p>
  </div>
"
`;

exports[`listCasesRoute > returns a list of cases 1`] = `
"

  <div class="govuk-grid-row">
    <div class="govuk-grid-column-full">
      <div class="govuk-body" data-testid="app-page-body">

        <div class="govuk-tabs" data-module="govuk-tabs">
  <h2 class="govuk-tabs__title">
    Contents
  </h2>
  <ul class="govuk-tabs__list">
    <li class="govuk-tabs__list-item govuk-tabs__list-item--selected">
      <a class="govuk-tabs__tab" href="#all-cases">
        SFI applications (2)
      </a>
    </li>
  </ul>
  <div class="govuk-tabs__panel" id="all-cases">
    <form method="GET" action="/cases/assign-user">
      <fieldset class="govuk-fieldset">
        <legend class="govuk-fieldset__legend govuk-visually-hidden">
          <h1>Cases</h1>
        </legend>
    
        <button type="submit" class="govuk-button" data-module="govuk-button">
      Assign
    </button>
    
        <table class="govuk-table">
      <thead class="govuk-table__head">
        <tr class="govuk-table__row">
          <th scope="col" class="govuk-table__header">Select</th>
          <th scope="col" class="govuk-table__header">ID</th>
          <th scope="col" class="govuk-table__header">Business</th>
          <th scope="col" class="govuk-table__header">SBI</th>
          <th scope="col" class="govuk-table__header">Submitted</th>
          <th scope="col" class="govuk-table__header">Status</th>
          <th scope="col" class="govuk-table__header">Assignee</th>
        </tr>
      </thead>
      <tbody class="govuk-table__body">
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">  <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="caseId-68495db5afe2d27b09b2ee47" name="caseId" type="radio" value="68495db5afe2d27b09b2ee47" aria-label="Select case ID 68495db5afe2d27b09b2ee47"><label class="govuk-label govuk-radios__label" for="caseId-68495db5afe2d27b09b2ee47">
          <span class="govuk-visually-hidden">
            Select case ID 68495db5afe2d27b09b2ee47
          </span>
        </label>
      </div>
    </td>
          <td class="govuk-table__cell">  <a href="/cases/68495db5afe2d27b09b2ee47" class="govuk-link govuk-!-font-weight-bold">banana-123</a>
    </td>
          <td class="govuk-table__cell">[business name]</td>
          <td class="govuk-table__cell">SBI001</td>
          <td class="govuk-table__cell">01 Oct 2023</td>
          <td class="govuk-table__cell"><strong class="govuk-tag govuk-tag--blue">
      New
    </strong>
    </td>
          <td class="govuk-table__cell">John Smith</td>
        </tr>
        <tr class="govuk-table__row">
          <td class="govuk-table__cell">  <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="caseId-68495db5afe2d27b09b2ee99" name="caseId" type="radio" value="68495db5afe2d27b09b2ee99" aria-label="Select case ID 68495db5afe2d27b09b2ee99"><label class="govuk-label govuk-radios__label" for="caseId-68495db5afe2d27b09b2ee99">
          <span class="govuk-visually-hidden">
            Select case ID 68495db5afe2d27b09b2ee99
          </span>
        </label>
      </div>
    </td>
          <td class="govuk-table__cell">  <a href="/cases/68495db5afe2d27b09b2ee99" class="govuk-link govuk-!-font-weight-bold">banana-123</a>
    </td>
          <td class="govuk-table__cell">[business name]</td>
          <td class="govuk-table__cell">SBI001</td>
          <td class="govuk-table__cell">01 Oct 2023</td>
          <td class="govuk-table__cell"><strong class="govuk-tag govuk-tag--grey">
      Old
    </strong>
    </td>
          <td class="govuk-table__cell">Not assigned</td>
        </tr>
      </tbody>
    </table>
    
      </fieldset>
    </form>
  </div>
</div>

      </div>
    </div>
  </div>
        "
`;
